File: 208-ImplementTrie.java
/**
 * 
 * Time Complexity : O(1) to search.
 * Space Complexity : 
 * Did this code successfully run on Leetcode : Yes
 * Any problem you faced while coding this : No
 * 
 Tries are DS to store strings
 They have same search complexity as hashmap but it is an optimal DS since it takes lesser space than a hashmap.
 Whenever a question asks to search entire word, prefix or suffix, always go with Trie first.
 In a trie, the root node contains all 26 chars(if only lowercase allowed..it would be 52 if uppercase are also allowed)
 now each child further has its own 26 chars' children.
 In order to know when the string end, at each trieNode we also store a boolean isEnd.
 To know the index of the current character we do [character - 'a']
 
 */

class Trie {

    class TrieNode{
        
        boolean isEnd;
        TrieNode[] children;
        public TrieNode(){
            
            children = new TrieNode[26];
        }
    }
    
    TrieNode root;
    /** Initialize your data structure here. */
    public Trie() {
        
        root = new TrieNode();
        
        
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            
            char c = word.charAt(i);
            if(curr.children[c - 'a'] == null){
                curr.children[c - 'a'] = new TrieNode();
            }
            curr = curr.children[c - 'a'];
        }
        curr.isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            
            char c = word.charAt(i);
            if(curr.children[c - 'a'] != null) {
                
                curr = curr.children[c - 'a'];
            }
            else return false;
            
        }
        return curr.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        
        TrieNode curr = root;
        for(int i = 0; i < prefix.length(); i++){
            
            char c = prefix.charAt(i);
            if(curr.children[c - 'a'] != null) {
                
                curr = curr.children[c - 'a'];
            }
            else return false;
            
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208_implementTrie.cpp
// Time Complexity : O(L) where L is the length of the word
// Space Complexity : O(N*K) where N is length of word and K is space used for building the Trie
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Trie {
public:
    class TrieNode{
        public:
        vector<TrieNode*> child;
        bool isEnd;
        TrieNode(){
            child = vector<TrieNode*>(26, NULL) ;
            isEnd = false;
        }
    };
    TrieNode* root;
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* curr = root;
        for(int i=0; i<word.size(); i++){
            char c = word[i];
            if(curr->child[c - 'a'] == NULL){
                curr->child[c - 'a'] = new TrieNode();
            }
            curr = curr->child[c - 'a'];
        }
        curr->isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* curr = root;
        for(int i=0; i<word.length(); i++){
            char c = word[i];
            if(curr->child[c - 'a'] == NULL)
                return false;
            curr = curr->child[c - 'a'];
        }
        return curr->isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* curr = root;
        for(int i=0; i<prefix.length(); i++){
            char c = prefix[i];
            if(curr->child[c - 'a'] == NULL)
                return false;
            curr = curr->child[c - 'a'];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
